### PR TITLE
dhcp: delay restart after NAKs

### PR DESCRIPTION
**Pull request by bengal**
Hi,

at some point my DSL gateway's DHCP server started to NAK all requests
and this generated an endless loop of requests and responses which
flooded my LAN:

  systemd-networkd[97]: DHCP CLIENT (0x6b82b407): DISCOVER
  systemd-networkd[97]: DHCP CLIENT (0x6b82b407): OFFER
  systemd-networkd[97]: DHCP CLIENT (0x6b82b407): REQUEST (requesting)
  systemd-networkd[97]: DHCP CLIENT (0x6b82b407): NAK: requested address not available
  systemd-networkd[97]: DHCP CLIENT (0xea1390cb): REBOOTED
  systemd-networkd[97]: DHCP CLIENT (0xea1390cb): DISCOVER
  systemd-networkd[97]: DHCP CLIENT (0xea1390cb): OFFER
  systemd-networkd[97]: DHCP CLIENT (0xea1390cb): REQUEST (requesting)
  systemd-networkd[97]: DHCP CLIENT (0xea1390cb): NAK: requested address not available
  systemd-networkd[97]: DHCP CLIENT (0xf68000fd): REBOOTED
  systemd-networkd[97]: DHCP CLIENT (0xf68000fd): DISCOVER
  systemd-networkd[97]: DHCP CLIENT (0xf68000fd): OFFER
  systemd-networkd[97]: DHCP CLIENT (0xf68000fd): REQUEST (requesting)
  systemd-networkd[97]: DHCP CLIENT (0xf68000fd): NAK: requested address not available
  systemd-networkd[97]: DHCP CLIENT (0xb334e56c): REBOOTED

This is probably caused by some misconfiguration on server side (most
likely a conflict between dynamic and static leases); however I think
that the client should handle better this situation.

For example, it seems that dhclient ignores NAKs and keeps sending
REQUESTs at the usual intervals; which is not strictly RFC-compliant,
but at least it avoids the mentioned issue.

This patch tries to improve networkd's behavior but probably there are
also other better approaches. Any suggestions?

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -952,10 +956,16 @@ static int client_initialize_time_events(sd_dhcp_client *client) {
 
         client->timeout_resend = sd_event_source_unref(client->timeout_resend);
 
+        if (client->init_delay) {
+                sd_event_now(client->event, clock_boottime_or_monotonic(), &usec);
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c:960]: assert_se(... >= 0);

**Comment by keszybz** [comment]
What does the standard say about this? Shouldn't the delay increase exponentially until some limit (1h?)?

You define two variables, the `nak_count` and `init_delay`. It should be enough to drop `nak_count` and increase `init_delay` wherever you currently increase `nak_count`, and reset `init_delay` where you reset `nak_count`. Implementing exponential falloff would be easy then.

**Comment by bengal** [comment]
> What does the standard say about this? Shouldn't the delay increase exponentially until some limit (1h?)

The RFC is not very clear about it (see commit message), it only says
that the configuration process must be restarted. The random initial
delay between 1 and 10 seconds is only a (maybe forced) interpretation
of the standard. But also increasing exponentially the delay would be
an option I guess.

>  It should be enough to drop nak_count and increase init_delay

Yeah, I'll try that, thanks.

**Comment by keszybz** [comment]
Yeah, the RFC is not very helpful. But I think we should implement a bigger falloff. Random 1-10s delay is an average of 15k attempts per day. It's rather pointless and floods the logs.

@teg ?

**Comment by poettering** [comment]
It appears to be this should be some exponential backoff logic here. double the interval each time, and top it up at 1h or so...

**Pull request closed without merging**